lib/linklist.c

```diff
@@ -92,6 +92,52 @@ void listnode_add_head(struct list *list, void *val)
 	list->count++;
 }
 
+bool listnode_add_sort_nodup(struct list *list, void *val)
+{
+	struct listnode *n;
+	struct listnode *new;
+	int             ret;
+
+	assert(val != NULL);
+
+	if (list->cmp) {
+		for (n = list->head; n; n = n->next) {
+			ret = (*list->cmp)(val, n->data);
+			if (ret < 0) {
+				new = listnode_new();
+				new->data = val;
+
+				new->next = n;
+				new->prev = n->prev;
+
+				if (n->prev)
+					n->prev->next = new;
+				else
+					list->head = new;
+				n->prev = new;
+				list->count++;
+				return true;
+			}
+			/* found duplicate return false */
+			if (ret == 0)
+				return false;
+		}
+	}
+
+	new = listnode_new();
+	new->data = val;
+	new->prev = list->tail;
+
+	if (list->tail)
+		list->tail->next = new;
+	else
+		list->head = new;
+
+	list->tail = new;
+	list->count++;
```

Comment:
I think we can use `LISTNODE_ATTACH(list, new)` to replace these lines?

Reply:
Yes. We can replace with the macro.